### PR TITLE
Upgrade to Mapnik 3.5

### LINKED ISSUE #109: Upgrade to Mapnik 3.5.0
https://github.com/mapnik/node-mapnik/blob/master/CHANGELOG.md#350

### PR DESCRIPTION
**Pull request by JesseCrocker**
closes #109  

Mapnik 3.5 compatibility ended up only being a 1 line change, but after upgrading 2 tests failed. Test results:
```
      ✓ should render a PNG of the world
      ✓ should render a PNG of Hispaniola
    #renderToVector()
      1) should render a GeoJSON
      2) should render a PBF

  #tree()
    ✓ should retrieve dir tree


  9 passing (462ms)
  2 failing

  1) #Tile() #renderToVector() should render a GeoJSON:

      Uncaught AssertionError: { type: 'FeatureCollection',
  features:
   [ { type: 'Feature',
       id: 45,
       geometry: [Object],
       properties: [ deepEqual { type: 'FeatureCollection',
  features:
   [ { type: 'Feature',
       id: 45,
       geometry: [Object],
       properties: [
      + expected - actual

                     19.915257
                   ]
                   [
                     -73.300781
      -              19.778334
      +              19.779627
                   ]
                   [
                     -73.300781
                     19.611544

      at compareGeoJSON (test/tile.js:19:19)
      at test/tile.js:67:17

  2) #Tile() #renderToVector() should render a PBF:

      Uncaught AssertionError: <Buffer 1a ff 01 78 02 0a 05 77 6f 72 6c 64 28 80 20 1a 02 69 64 22 05 0a 03 44 4f 4d 1a 04 6e 61 6d 65 22 14 0a 12 44 6f 6d 69  deepEqual <Buffer 1a ff 01 78 02 0a 05 77 6f 72 6c 64 28 80 20 1a 02 69 64 22 05 0a 03 44 4f 4d 1a 04 6e 61 6d 65 22 14 0a 12 44 6f 6d 69
      + expected - actual

         5
         197
         1
         0
      -  129
      +  131
         2
         162
         1
      -  211
      +  209
         1
         15
         8
         71

      at test/tile.js:76:24
```

These failures both seemed really minor, and most likely caused by changes in how mapnik is doing simplification. The new results seemed as valid as the old results, so I fixed it by updating the expected test results.

### CODE REVIEW

**Comment by yohanboniface** [comment]
Yeah, those tests are a bit annoying, they break every time we upgrade Mapnik. cf #111 

Strangely, there is still one run failing on Travis, any clue why?

Anyway, thanks a lot for working at this :)

**Comment by JesseCrocker** [comment]
It looks like that failing test is an error installing mapnik, not a test failure.

**Comment by yohanboniface** [comment]
But is that in a way linked to the 3.5 version of Mapnik?

**Comment by yohanboniface** [comment]
@springmeyer, does that ring a bell to you?

https://travis-ci.org/kosmtik/kosmtik/jobs/119282535

```
gyp ERR! UNCAUGHT EXCEPTION 

gyp ERR! stack Error: Cannot find module 'rimraf'
```

It's with Node 0.12 only.

**Comment by springmeyer** [comment]
@yohanboniface as of node-mapnik >= 3.5.3 we only provide pre-built binaries for node v0.10, v4, and v5. I wish there were a way to make this more clear. That error is a really really odd result for binaries being missing. I wonder if it has something to do with how we bundle node-pre-gyp (https://github.com/mapbox/node-pre-gyp/issues/162). Either way the fix for you is to stop testing on node v0.12. If this is a major problem let me know, but I made the change with the presumption that anyone that was once using node v0.12 has upgraded to v4 or v5.

**Comment by springmeyer** [comment]
> Yeah, those tests are a bit annoying, they break every time we upgrade Mapnik. cf #111

If these tests are comparing floating point values and/or generated from vector tiles then changing per node-mapnik release is expected and 99% of the time a really good thing. There has been much work on improving vector tile geometry output and robustness which is bound to change floating point values.

File: src/back/VectorBasedTile.js [hunk 1]
```diff
@@ -18,7 +18,6 @@ VectorBasedTile.prototype._render = function (project, map, cb) {
         parse = function (data, resp) {
             try {
                 vtile.setData(data);
```
Comment (springmeyer) [inline src/back/VectorBasedTile.js]: In node-mapnik >= 3.5 `setData` does more work that previously. So I would recommend moving to the async version of `setData` for best performance. Before it did not matter. /cc @yohanboniface

**Comment by springmeyer** [comment]
Thanks for knocking this out @JesseCrocker. I'll add that if any other issues come up during the upgrade feel free to ping or look through the list of pulls tallied at https://github.com/mapbox/mapnik-swoop/pull/10 (which you can see what it took to upgrade a variety of other modules).

**Comment by JesseCrocker** [comment]
I've been using this branch a bunch yesterday and today for vector tile based projects, and everything seems to be working. I also did some testing with non-vectortile project, and that also worked.

**Comment by yohanboniface** [comment]
Quickly tested on some projects I have, it brokes on vector ones, an example is https://github.com/yohanboniface/thank-you-map
I'll try to dig in but today I'm a bit in a rush, and I'm leaving tomorrow morning to Madrid for Leaflet sprint, so I may be late on this :s

**Comment by yohanboniface** [comment]
Also, I think we should warn on README that latest master (and future npm release) will work only with Node 0.10, 4.x and 5.x, what do you think?

**Comment by yohanboniface** [comment]
For the record, traceback:
```
Vector Tile Validity Errors Found:
 - Vector Tile Layer message has no features

Trace
    at ProjectServer.raise (/home/ybon/Code/js/kosmtik/src/back/ProjectServer.js:239:13)
    at /home/ybon/Code/js/kosmtik/src/back/ProjectServer.js:73:34
    at /home/ybon/Code/js/kosmtik/src/back/VectorBasedTile.js:61:18
    at /home/ybon/Code/js/kosmtik/src/back/VectorBasedTile.js:22:28
Unable to parse vector tile data for uri http://vector.mapzen.com/osm/all/17/66389/45094.mapbox
```

**Comment by springmeyer** [comment]
The vector tile v1 spec says `A vector tile can consist of one or more named layers and containing one or more features.` at https://github.com/mapbox/vector-tile-spec/tree/master/1.0.1#3-internal-structure. So basically this is catching a layer that is empty which is pointless and the layer itself then should not have been encoded. In node-mapnik 3.5.x this is enforced so that we can easily know how to upgrade to v2 version tiles. So, the possible solutions are:

  - 1) Mapzen upstream fixes the tiles to not encode empty layers
  - 2) You change your tests to use other tiles that don't contain empty layers
  - 3) We add a mode to node-mapnik to not be so strict during decoding. Basically you'd get a warning on layers with no features instead of an error.

2) seems like the easiest thing to me and what I would recommend. Sound good?

/cc @flippmoke for gut check on my reasoning.

**Comment by flippmoke** [comment]
@springmeyer when I originally designed it I was planning on putting a warning and an error setting but due to time just set it to both currently. I should make it warn rather then error when the specification specifically says SHOULD rather then MUST.

**Comment by pnorman** [comment]
> @springmeyer when I originally designed it I was planning on putting a warning and an error setting but due to time just set it to both currently. I should make it warn rather then error when the specification specifically says SHOULD rather then MUST.

Given that v1 isn't explict on ruling out a zero-feature layer and v2 allow it, even if layers SHOULD contain features, I'd agree that it shouldn't error when parsing a v1 mvt with a zero-feature layer.

**Comment by yohanboniface** [comment]
I'd go for the [Robustness principle](https://en.wikipedia.org/wiki/Robustness_principle), then do not raise an error for an empty layer.

@springmeyer @flippmoke if you agree on that, any chance to get a new node-mapnik going this way? I think that would help the community upgrade, including us ;)

Thanks for your support, once again :)

**Comment by springmeyer** [comment]
> any chance to get a new node-mapnik going this way

@yohanboniface - in process: https://github.com/mapnik/node-mapnik/pull/626

**Comment by yohanboniface** [comment]
Excellent! Thank you :)

**Comment by JesseCrocker** [comment]
@yohanboniface new release of node-mapnik is out that should be compatible with those mapzen tiles.  https://github.com/mapnik/node-mapnik/releases/tag/3.5.9

**Comment by yohanboniface** [comment]
Much better!
But I still see some errors on some tiles:

`Vector Tile has LINESTRING type geometry with a MOVETO command with no LINETO following.`

`Vector Tile has POLYGON with first ring clockwise. It is not valid according to v2 of VT spec.`

I guess those are actual errors and need to be fixed on the vector provider side @springmeyer?

I still insist a bit on being liberal on what we accept, printing warnings to let the user fix those issues. But here I'm not sure if those are actual errors that prevent Mapnik from processing the tile or only issues with the specs. If it's the later, I think we should go with warning only.

**Comment by JesseCrocker** [comment]
I ran into that same error with one of my tile sets that i was generating with mapzens vector tile library https://github.com/mapzen/mapbox-vector-tile, The problem ended up not being LineStrings with a  single point as the documents suggest, but LineStrings with 2 points, but the points were coincident after converting to tile coordinates.

**Comment by yohanboniface** [comment]
> The problem ended up not being LineStrings with a single point as the documents suggest, but LineStrings with 2 points, but the points were coincident after converting to tile coordinates.

Seems something to be fixed in the proto definition (https://github.com/mapzen/mapbox-vector-tile/tree/master/mapbox_vector_tile/Mapbox). But again I'm tempted to think that Mapnik should be more tolerant here. Or we will run in a lot of people having their setup broken without being able to fix it (the provider needs to fix it).

> I ran into that same error with one of my tile sets that i was generating with mapzens vector tile library https://github.com/mapzen/mapbox-vector-tile

BTW, you may want to have a look at [Utilery](http://utilery.readthedocs.org/) for an easiest path. ;)

**Comment by springmeyer** [comment]
> I guess those are actual errors and need to be fixed on the vector provider side @springmeyer?

Unless there is a mapnik-vector-tile bug that is incorrectly parsing the `version` from those tiles (which is very unlikely) those errors indicate the tiles are claiming to be v2. So, they need to be fixed to be valid `v2` by the provider or you need to use a `v1` endpoint (which latest node-mapnik will handle gracefully).

**Comment by yohanboniface** [comment]
Good point, let me check that so we are sure.
Is there any one liner to validate a pbf tile from an URL?

`mapnik-validate http://domain.com/tile.pbf` :)

**Comment by springmeyer** [comment]
@yohanboniface - if you write something you can use the `mapnik.VectorTile.info` API:  http://mapnik.org/node-mapnik/documentation/3.5/#VectorTile.info

**Comment by yohanboniface** [comment]
Seems like mapbox-vector-tile is pretending to return V2 tiles, indeed:

```
{ layers: 
   [ { name: 'boundaries',
       features: 4,
       point_features: 0,
       linestring_features: 4,
       polygon_features: 0,
       unknown_features: 0,
       raster_features: 0,
       version: 2 },
     { name: 'buildings',
       features: 167,
       point_features: 74,
       linestring_features: 0,
       polygon_features: 93,
       unknown_features: 0,
       raster_features: 0,
       version: 2 },
 …
```


See also https://github.com/mapzen/mapbox-vector-tile/issues/42#issuecomment-210141373

**Comment by springmeyer** [comment]
> For reference, the script I've been using for getting the info from an URL: https://gist.github.com/yohanboniface/e8f493e96d42cc3d154631b35b836772

@yohanboniface fwiw `mapnik.VectorTile.info` takes both zlib and gzip compressed buffers so you don't have to detect that yourself or decompress beforehand.

**Comment by yohanboniface** [comment]
OK, let's move on. Thanks @JesseCrocker for the work, and @springmeyer as usual for the kind help :)

**Pull request merged**